src/hasPropSatisfying.test.ts

```diff
@@ -0,0 +1,249 @@
+  });
+
+  it('should work when using another hasPropSatisfying as the predicate', () => {
+    const obj = { data: { foo: 'bar' } };
+    expect(
+      hasPropSatisfying(obj, 'data', hasPropSatisfying('foo', isEqual('bar')))
+    ).toBe(true);
+  });
+
+  describe('predicate argument type narrowing', () => {
+    it('should narrow the type of the predicate argument to exclude undefined', () => {
+      const obj = { a: undefined } as { a: number | undefined };
+      hasPropSatisfying(obj, 'a', (val): val is 1 => {
+        expectTypeOf(val).toEqualTypeOf<number>();
+        return true;
+      });
+    });
+
+    it('should narrow the type of the predicate argument to exclude undefined when the prop is optional', () => {
+      const obj = { a: undefined } as { a?: number | undefined };
+      hasPropSatisfying(obj, 'a', (val): val is 1 => {
+        expectTypeOf(val).toEqualTypeOf<number>();
+        return true;
+      });
+    });
+  });
+
+  describe('type narrowing', () => {
+    it('should narrow the type of the specified prop based on the provided predicate', () => {
+      const obj = { a: 1 } as { a?: number | string };
+      if (hasPropSatisfying(obj, 'a', isEqual(1))) {
+        expectTypeOf(obj).toEqualTypeOf<{ a: 1 }>();
+      }
+      if (hasPropSatisfying(obj, 'a', isNumber)) {
+        expectTypeOf(obj).toEqualTypeOf<{ a: number }>();
+      }
+    });
+
+    it('should narrow the type of an object union based on the predicate result', () => {
+      const obj = { type: 'success', data: { foo: 'bar' } } as
+        | { type: 'success'; data: { foo: string } }
+        | { type: 'error'; error: string };
+
+      if (hasPropSatisfying(obj, 'type', isEqual('success'))) {
+        expectTypeOf(obj).toEqualTypeOf<{
+          type: 'success';
+          data: { foo: string };
+        }>();
+      }
+
+      if (hasPropSatisfying(obj, 'type', isEqual('error'))) {
+        expectTypeOf(obj).toEqualTypeOf<{ type: 'error'; error: string }>();
+      }
+    });
+
+    it('should narrow the type of an object union based on property presence', () => {
+      const obj = { type: 'success', data: { foo: 'bar' } } as
+        | { type: 'success'; data: { foo: string } }
+        | { type: 'error'; error: string };
+
+      if (hasPropSatisfying(obj, 'data', isObjectType)) {
+        expectTypeOf(obj).toEqualTypeOf<{
+          type: 'success';
+          data: { foo: string };
+        }>();
+      }
+    });
+
+    it('should be able to accept another hasPropSatisfying as a predicate', () => {
+      const obj = { type: 'success', data: { foo: 'bar' } } as
+        | { type: 'success'; data: { foo: string } }
+        | { type: 'error'; error: string };
+
+      if (
+        hasPropSatisfying(obj, 'data', hasPropSatisfying('foo', isEqual('bar')))
+      ) {
+        expectTypeOf(obj).toEqualTypeOf<{
+          type: 'success';
+          data: { foo: string } & { foo: 'bar' };
```

Comment:
When nesting the helper like this, the type doesn't simplify quite as well. Using `SimplifyDeep` from the typefest files fixes it, but I'm not so sure we should use that for this helper.

I imagine using `Omit` to remove the old version of the value would fix this, but that causes issues in other places as I mentioned in other comments.

But I suspect this is probably an uncommon use-case, so it seems like a reasonable compromise.

Reply:
After doing some more research, I figured out how to properly address this. So now we're able to get the correct, fully resolved type.